### PR TITLE
ARROW-8906: [Rust] [DataFusion] support schema inference from multiple CSV files

### PR DESCRIPTION
**Pull request by houqp**
This change adds `infer_schema_from_files` function to arrow csv reader module. Datafusion's `CsvExec` struct is now using this function to do schema inference from multiple CSV files if needed.

Second follow up PR for #7210.

### CODE REVIEW

File: rust/arrow/src/csv/reader.rs
Comment (andygrove) [inline rust/arrow/src/csv/reader.rs:191]: I'm a little nervous about the idea of potentially loading all of the csv files into memory. I'm also not sure how this will work when the schema varies between files. In the unit test, all files appear to have the same three columns. What if one of the files is missing "c2" and another file has an additional "c4" and "c5" ?

What I think we'll ultimately need in DataFusion is a schema merging feature, so each csv (or parquet) file can have differences but the final output schema will contain the superset.

Comment (andygrove) [reply rust/arrow/src/csv/reader.rs:191]: Another option here is that this code returns an error if the files have differing schemas.

Comment (houqp) [reply rust/arrow/src/csv/reader.rs:191]: sounds good, i will change it to infer schema per file, then merge at the end. if we don't have to seek across multiple files, then we don't need to load all lines into memory.

**Comment by houqp** [comment]
@andygrove I have added schema merge logic to `Schema` and `Field` structs. As a result, we are now able to do schema inference per file thus avoiding the need to load records into memory.

File: rust/arrow/src/csv/reader.rs
Comment (nevi-me) [inline rust/arrow/src/csv/reader.rs:209]: Just checking, we can never get `records_to_read` < 0, right?

**Review by nevi-me — COMMENTED** [review]
Just leaving a minor note, I'm left with looking at the schema merging functions.

File: rust/arrow/src/csv/reader.rs
Comment (houqp) [reply rust/arrow/src/csv/reader.rs:209]: That's correct, should always be positive as long as infer_file_schema doesn't read more records than requested. I used `<=` below to exit early if any unexpected error occurs. We can also throw an error when it's less than zero.

File: rust/arrow/src/datatypes.rs
Comment (nevi-me) [inline rust/arrow/src/datatypes.rs:1420]: May you please add a doc for this

**Review by nevi-me — APPROVED** [review]
LGTM

File: rust/arrow/src/datatypes.rs
Comment (houqp) [reply rust/arrow/src/datatypes.rs:1420]: ha, good catch, i added a doc for this, but it was mistakenly added to Field's try_merge method. I have fixed and also added doc for Field's try_merge method as well.

Comment (nevi-me) [reply rust/arrow/src/datatypes.rs:1431]: Thanks, I'll await a review from another commiter, then we can merge

**Comment by nevi-me** [comment]
I'll merge this tomorrow evening GMT if there haven't been any further reviews.

**Pull request closed without merging**
